### ISSUE TITLE
Anomaly in nontrivial unification in 8.10beta1

### ISSUE THREAD

**Issue by charguer** [coq/coq#10285]
#### Description of the problem

I tried by best to minimize the instance, but further simplifications get the anomaly to turn into an (expected) type error. 

```coq
Parameter heap : Type.

Definition hprop := heap -> Prop.

Definition pred_incl (A : Type) (P Q : A -> Prop) :=
  forall x, P x -> Q x.

Definition hdata A (P:A->hprop) (X:A) : hprop := P X.

Parameter Ref : forall A, A -> nat -> hprop.

Lemma test : forall
  (A : Type)
  (r : nat)
  (H : forall (H : hprop) (Q : unit -> hprop), 
       Q = (fun _ : unit => H) -> @pred_incl heap H (Q tt))
  (X: forall B,
       @pred_incl heap (@hdata nat (@Ref (list B) (@nil B)) r)
                       (@hdata nat (@Ref (list A) (@nil A)) r) 
   -> True),
  True.
Proof using.
  intros. eapply X. clear X. apply H.
Qed.
```

#### Coq Version

8.10beta1 fails.

The anomaly does not show up in 8.8.2 nor 8.9.0. The expected error messages looks like:
`Unable to unify "?M163 tt x" with  "Ref (list A) nil r x".`

**Comment by ejgallego** [comment]
Looks like an error when trying to print the error message, but not sure cc: @herbelin

**Comment by SkySkimmer** [comment]
I think it's failing to find some evar at evarutil line 527.

**Comment by ejgallego** [comment]
> I think it's failing to find some evar at evarutil line 527.

That's the trace I saw indeed.

**Comment by ejgallego** [comment]
But note @SkySkimmer that if you put the example in a file `foo.v` and run `coqc foo.v` you get no anomaly, so that makes the printing code suspicious?

**Comment by ejgallego** [comment]
This is due to #7819 , however the rebase couldn't pinpoint the exact problem as none of these commits do build

```
There are only 'skip'ped commits left to test.
The first bad commit could be any of:
1c60cbedfd8a5e64bfa95dfb9a9497e278458c30
66892881c69cfb91fe1cd772cd88ac2c73c99bbe
78b51f541d0107f06c21fc1260aae2ab9f7229c5
8a5cd826ee8d7d51c85c834c28f090466dda33a5
a4157eb4cb5ede453e02b415aa0c2b10ce9f961d
acc4f5922dcc1f92f3dc3db653b7608949b60c2b
93ef4e058cb9f7bfc6f3abc8bdc5752a2d8df5ca
c002eae68ac12dc8ed92e906b346f73606e7fd69
2cf15ef8e3ddebe285ec0e9f5f8715b025ba4eec
e3178fb2de9a84bbcdc90a60cd8f27cd1323eb36
af9b88c253b578ad53ee884ff3102d0a74056a1e
e07fcc19cf8459ce2ae87d9ccbbc7806a0be576f
2265d20c48dff6e1ea9a5cb9be9640603f3be3cf
5f05bb81c2de54d31887b276d0f127455dd83815
781f050bcfabe02e225f3c1d29ee649610d6d680
ad0cf24789b0f7d5b12cef4d09f93c9d9aeb6150
62cf1ad380772628ed6e547de3fb875275610b4d
dc8f191f2d81fd5f851b84e7aeda487df584197e
1c9e1a39652b401805029519055aa62adacde339
ce36c131804cd7fb0a460277d88743c7131c5ed3
746e74ed6e70cf01f0a0e73f772c3565e28fe3f8
b74aaa4ce0191b92837a15773ae23e53abd22bd9
ab6397730354d6048062568c87a7d971f20934c0
```

**Comment by ejgallego** [comment]
Using a fixed branch the bad commit is:
```
Author: Matthieu Sozeau <matthieu.sozeau@inria.fr>
Date:   Thu Jul 26 17:07:08 2018 +0200

    [evarconv] New flag handling for unifier
```
746e74ed6e70cf01f0a0e73f772c3565e28fe3f8

Maybe a duplicate of #10300 ?

**Comment by vbgl** [comment]
During a call to `thin_evars`, an evar is first found in the initial `sigma` (at https://github.com/coq/coq/blob/82aa7ed1c929533f2b9e7acaa8264c5a81e7a45e/pretyping/evarconv.ml#L1318) and then not found in the restricted `!evdref` (https://github.com/coq/coq/blob/82aa7ed1c929533f2b9e7acaa8264c5a81e7a45e/pretyping/evarconv.ml#L1322).

**Comment by vbgl** [comment]
This has been fixed in 8.10 by 43e27412cb7e349d372b28470e11ef755d8d762a.
